### DISCUSSION TITLE
Interactively select/pick a row in a table?

### DISCUSSION THREAD

**Discussion by daniel-shimon** [Textualize/rich#960] [Q&A] [answered]
Hi!

Is it possible to achieve something like this with rich?

Or possibly extend rich somehow to achieve it?

**Comment by willmcgugan** [comment, marked as answer]
Not currently, but maybe in the future. In the meantime you could try PyInquirer.

**Reply by akshaybabloo** [reply]
PyInquirer is good but I don't think it's being maintained. See https://github.com/CITGuru/PyInquirer/issues/159. Also, their use of `prompt_toolkit==1.0.14` will break other CLIs

**Reply by yashrathi-git** [reply]
@akshaybabloo yes, I too have problems of it clashing with JupyterNotebook. Try [questionary](https://github.com/tmbo/questionary). It have: `prompt_toolkit = ">=2.0,<4.0"`. And is really good and well-maintained!

**Comment by Spenhouet** [comment]
@willmcgugan Having an interactive prompt would still be very much appreciated.

There seem to be many feature requests on this, for example:

- https://github.com/willmcgugan/rich/issues/498
- https://github.com/willmcgugan/rich/issues/262

Other projects with this functionality:

- https://github.com/CITGuru/PyInquirer
- https://github.com/manifoldco/promptui
- https://github.com/IngoMeyer441/simple-term-menu
- https://github.com/magmax/python-inquirer

Will this feature come?

**Reply by willmcgugan** [reply]
Not in Rich, but it will in Textual.

**Reply by Spenhouet** [reply]
I wasn't aware of Textual but it looks a bit overkill for us. We would like to combine typer with rich. So I guess there will not be an interactive selection prompt.
A simple implementation would just show a numbered list and allow to enter the number (that is how we have implemented it for now). Often the options are just too long for the user to type it down. Especially if they are already fixed / known. This could be combined with an enum manage identifiers and names.

**Reply by willmcgugan** [reply]
The implementation would not be much different if it were added to Rich, or Textual. Textual uses Rich as a renderer, so you can consider it interactive Rich. But that level of interactivity is no small amount of work, with drivers for both Linux and Windows, so it won't be in both libraries.

Should you need it for a commercial product, you might want to consider [sponsoring the work](https://github.com/sponsors/willmcgugan).

**Reply by Spenhouet** [reply]
Does Textual work together with typer?

**Reply by willmcgugan** [reply]
There's nothing that would prevent them working together. They do quite different things.

**Reply by Spenhouet** [reply]
Okay, thanks. I did not look into it.
I did just see the example and it looked incompatible with typer:

Textual example:

```python
from textual.app import App

class Beeper(App):
    def on_key(self):
        self.console.bell()

Beeper.run()
```

typer example:

```python
import typer

app = typer.Typer()


@app.command()
def hello(name: str):
    typer.echo(f"Hello {name}")


@app.command()
def goodbye(name: str, formal: bool = False):
    if formal:
        typer.echo(f"Goodbye Ms. {name}. Have a good day.")
    else:
        typer.echo(f"Bye {name}!")


if __name__ == "__main__":
    app()
```

But I guess I would need to check it out.

**Reply by willmcgugan** [reply]
You wouldn't do it that way. You can call Beeper.run from any of the commands to gather the information you need. But bear in mind Textual is a work in progress, unlike Rich which is relatively mature.

**Comment by ppmx** [comment]
Brewing this thread up again... I wrote a small PoC for an interactive selector app using textual.

```py
from textual.app import App
from textual.widget import Widget
from rich.panel import Panel

class Hover(Widget):
    def __init__(self, label, *args, **kwargs):
        self.label = label
        self._is_active = False
        super().__init__(*args, **kwargs)

    def render(self) -> Panel:
        return Panel(self.label, style=("on red" if self.is_active else ""))

    @property
    def is_active(self) -> None:
        return self._is_active

    @is_active.setter
    def is_active(self, new_state) -> None:
        self._is_active = new_state
        self.refresh()

class TextualChoiceSelector(App):
    def __init__(self, choices, *args, **kwargs):
        self.choices = [Hover(choice) for choice in choices]
        self.active_choice = 0
        self.choices[self.active_choice].is_active = True

        super().__init__(*args, **kwargs)

    def on_key(self, event):
        if event.key in ["down", "up"]:
            off = 1 if event.key == "down" else -1

            self.choices[self.active_choice].is_active = False
            self.active_choice = (self.active_choice + off) % len(self.choices)
            self.choices[self.active_choice].is_active = True

        if event.key == "enter":
            pass #...

    async def on_mount(self) -> None:
        await self.view.dock(*self.choices, edge="top")

TextualChoiceSelector.run(choices=["Select A", "Select B", "Select C", "Select D", "Select E"])
```

**Reply by glensc** [reply]
@ppmx is the code incomplete? or already outdated?

```
$ python test_rich_input.py
Traceback (most recent call last):
  File "test_rich_input.py", line 45, in <module>
    TextualChoiceSelector.run(choices=["Select A", "Select B", "Select C", "Select D", "Select E"])
TypeError: App.run() got an unexpected keyword argument 'choices'
```

```
Python 3.12.1
rich               13.7.0
textual            0.47.1
```

**Reply by glensc** [reply]
Seems it's outdated, `def dock` got removed in:
- https://github.com/Textualize/textual/commit/b3eb543e380c904768b59c051e2415e233d8516f

which went into 0.2.0 release:
- https://github.com/Textualize/textual/pull/660

```
$ pip install textual"<0.2.0"
```

now the snippet works.

**Comment by dsm-72** [comment]
> Not currently, but maybe in the future. In the meantime you could try PyInquirer.

@willmcgugan it is the future. I'd like to add my voice to those who would like `rich` to support this.

**Comment by noisy** [comment]
@daniel-shimon @Spenhouet @dsm-72 @ppmx 

I've started working on a [rich_interactive](https://github.com/noisy/rich_interactive) package , which extends `rich` classes and add interactive features.

![image](https://github.com/Textualize/rich/assets/201263/db4ea476-1e19-466d-9e50-1d32f1f2c296)
![image](https://raw.githubusercontent.com/noisy/rich_interactive/main/docs/images/interactive_layout.png)

I started just a few days ago, so this is very early stage, but if anyone would be interested in such a package, I would appreciate a feedback/ideas/requests.

My personal goal, is to prepare a good quality package, which would enable easy navigation with keyboard, like this:

![live demo](https://github.com/Textualize/rich/assets/201263/866657eb-2814-4d5a-8150-eb81069f172a)

(Disclaimer: gif with animated part is in PoC stage, and it is not production ready yet)

**Reply by NomadicDeveloper22** [reply]
Oh wow, this looks awesome!

**Comment by synacktraa** [comment]
I've a created a small library [minifzf](https://github.com/synacktraa/minifzf), a while ago, It might be useful.

This is how it looks like:
![image](https://github.com/Textualize/rich/assets/91981716/c244c300-7eac-41c8-ad31-2f28a8f08061)

It also has fuzzy search capability. Use `f` or `/` to enable search bar. I'm releasing a new version with more capabilities(multi-select and improved search-bar)...will update the documentation, it's empty :).
